signature-aggregator/aggregator/cache/cache.go

```diff
@@ -0,0 +1,69 @@
+package cache
+
+import (
+	"math"
+
+	"github.com/ava-labs/avalanchego/ids"
+	"github.com/ava-labs/avalanchego/utils/crypto/bls"
+	"github.com/ava-labs/avalanchego/utils/logging"
+	lru "github.com/hashicorp/golang-lru/v2"
+	"github.com/pingcap/errors"
+	"go.uber.org/zap"
+)
+
+type Cache struct {
+	logger     logging.Logger
+	signatures *lru.Cache[cacheKey, cacheValue]
+}
+
+type cacheKey ids.ID // warp message ID
+type cacheValue map[PublicKeyBytes]SignatureBytes
+
+type PublicKeyBytes [bls.PublicKeyLen]byte
+type SignatureBytes [bls.SignatureLen]byte
+
+func NewCache(size uint, logger logging.Logger) (*Cache, error) {
+	if size > math.MaxInt {
+		return nil, errors.New("cache size too big")
+	}
+
+	signatureCache, err := lru.New[cacheKey, cacheValue](int(size))
+	if err != nil {
+		return nil, err
+	}
+
+	return &Cache{
+		signatures: signatureCache,
+		logger:     logger,
+	}, nil
+}
+
+func (c *Cache) Get(msgID ids.ID) (cacheValue, bool) {
+	cachedValue, isCached := c.signatures.Get(cacheKey(msgID))
+
+	if isCached {
+		c.logger.Debug("cache hit", zap.Stringer("msgID", msgID))
+		return cachedValue, true
+	} else {
+		c.logger.Debug("cache miss", zap.Stringer("msgID", msgID))
+		return make(cacheValue), false
+	}
+}
+
+func (c *Cache) Add(
+	msgID ids.ID,
+	pubKey PublicKeyBytes,
+	signature SignatureBytes,
+) {
+	cachedValue, ok := c.Get(msgID)
+	if !ok {
+		c.signatures.Add(cacheKey(msgID), make(cacheValue))
+		cachedValue, ok = c.Get(msgID)
+		if !ok {
+			c.logger.Error(
+				"failed to create cache entry; this should never happen",
+			)
+		}
```

Comment:
```suggestion
		cachedValue = make(cacheValue)
		c.signatures.Add(cacheKey(msgID), make(cacheValue))
```

Reply:
I tried this but it seems like the map that's constructed here (`cachedValue = make(cacheValue)`) is copied into the cache via `Add()`, so that the subsequent assignment to a map entry only affects the instance in the function scope, not the one that was added to the cache.

I got rid of the "should never happen" weirdness in https://github.com/ava-labs/awm-relayer/pull/415/commits/b8862e7524de554015c9a0224036e17eb0634036, in exchange for duplicating the signature assignment line. let me know what you think.